File: stepik-union/src/test/java/com/jetbrains/tmp/learning/StudySerializationUtilsTest.java
package com.jetbrains.tmp.learning;

import org.stepik.core.serialization.StudySerializationUtils;
import org.stepik.core.serialization.StudyUnrecognizedFormatException;
import com.thoughtworks.xstream.XStream;
import org.jdom.Element;
import org.jdom.output.Format;
import org.jdom.output.XMLOutputter;
import org.junit.BeforeClass;
import org.junit.Test;
import org.stepik.core.StepikProjectManager;
import org.stepik.core.TestUtils;
import org.xml.sax.SAXException;

import javax.xml.parsers.ParserConfigurationException;
import java.io.IOException;

import static org.stepik.core.serialization.StudySerializationUtils.MAIN_ELEMENT;
import static java.text.MessageFormat.format;
import static org.junit.Assert.assertEquals;

/**
 * @author meanmail
 */
public class StudySerializationUtilsTest {
    private static final int SOURCES_COUNT = 4;
    private static XMLOutputter outputter;

    @BeforeClass
    public static void before() {
        outputter = new XMLOutputter();
        Format format = Format.getPrettyFormat().setLineSeparator("\n");
        outputter.setFormat(format);
    }

    private static Element readSourceXmlFile(int version, int index)
            throws IOException, SAXException, ParserConfigurationException {
        return TestUtils.readXmlFile(StudySerializationUtilsTest.class, format("version{0}_{1}.xml", version, index));
    }

    private static String readSourceTextFile(int version, int index) throws IOException {
        return TestUtils.readTextFile(StudySerializationUtilsTest.class, format("version{0}_{1}.xml", version, index));
    }

    @Test
    public void convertToSecondVersion() throws Exception {
        Element stateVersion1 = readSourceXmlFile(1, 4);
        Element stateVersion2 = StudySerializationUtils.convertToSecondVersion(stateVersion1);

        String actual = outputter.outputString(stateVersion2);
        String expected = readSourceTextFile(2, 4);

        assertEquals("#4", expected, actual);
    }

    @Test
    public void convertToThirdVersion() throws Exception {
        for (int i = 1; i <= SOURCES_COUNT; i++) {
            Element stateVersion2 = readSourceXmlFile(2, i);
            Element stateVersion3 = StudySerializationUtils.convertToThirdVersion(stateVersion2);

            String actual = outputter.outputString(stateVersion3);
            String expected = readSourceTextFile(3, i);

            assertEquals("#" + i, expected, actual);
        }
    }

    @Test(expected = StudyUnrecognizedFormatException.class)
    public void convertToThirdVersionWithException() throws Exception {
        Element stateVersion2 = TestUtils.readXmlFile(StudySerializationUtilsTest.class, "version2_0.xml");
        StudySerializationUtils.convertToThirdVersion(stateVersion2);
    }

    @Test
    public void convertToFourthVersion() throws Exception {
        for (int i = 1; i <= SOURCES_COUNT; i++) {
            Element stateVersion3 = readSourceXmlFile(3, i);
            Element stateVersion4 = StudySerializationUtils.convertToFourthVersion(stateVersion3);

            XStream xs = StepikProjectManager.getXStream();

            Object obj = xs.fromXML(outputter.outputString(stateVersion4.getChild(MAIN_ELEMENT)));

            String actual = xs.toXML(obj); //outputter.outputString(stateVersion4);
            String expected = readSourceTextFile(4, i);

            assertEquals("#" + i, expected, actual);
        }
    }
}

File: stepik-union/src/test/kotlin/org/stepik/core/StudySerializationUtilsTest.kt
package org.stepik.core

import org.jdom.Element
import org.jdom.output.Format
import org.jdom.output.XMLOutputter
import org.junit.Assert.assertEquals
import org.junit.BeforeClass
import org.junit.Test
import org.stepik.core.serialization.StudySerializationUtils
import org.stepik.core.serialization.StudySerializationUtils.MAIN_ELEMENT
import org.stepik.core.serialization.StudyUnrecognizedFormatException
import org.xml.sax.SAXException
import java.io.IOException
import java.text.MessageFormat.format
import javax.xml.parsers.ParserConfigurationException


class StudySerializationUtilsTest {

    @Test
    @Throws(Exception::class)
    fun convertToSecondVersion() {
        val stateVersion1 = readSourceXmlFile(1, 4)
        val stateVersion2 = StudySerializationUtils.convertToSecondVersion(stateVersion1)

        val actual = outputter!!.outputString(stateVersion2)
        val expected = readSourceTextFile(2, 4)

        assertEquals("#4", expected, actual)
    }

    @Test
    @Throws(Exception::class)
    fun convertToThirdVersion() {
        for (i in 1..SOURCES_COUNT) {
            val stateVersion2 = readSourceXmlFile(2, i)
            val stateVersion3 = StudySerializationUtils.convertToThirdVersion(stateVersion2)

            val actual = outputter!!.outputString(stateVersion3)
            val expected = readSourceTextFile(3, i)

            assertEquals("#" + i, expected, actual)
        }
    }

    @Test(expected = StudyUnrecognizedFormatException::class)
    @Throws(Exception::class)
    fun convertToThirdVersionWithException() {
        val stateVersion2 = TestUtils.readXmlFile(StudySerializationUtilsTest::class.java, "version2_0.xml")
        StudySerializationUtils.convertToThirdVersion(stateVersion2)
    }

    @Test
    @Throws(Exception::class)
    fun convertToFourthVersion() {
        for (i in 1..SOURCES_COUNT) {
            val stateVersion3 = readSourceXmlFile(3, i)
            val stateVersion4 = StudySerializationUtils.convertToFourthVersion(stateVersion3)

            val xs = StepikProjectManager.xStream

            val obj = xs.fromXML(outputter!!.outputString(stateVersion4.getChild(MAIN_ELEMENT)))

            val actual = xs.toXML(obj) //outputter.outputString(stateVersion4);
            val expected = readSourceTextFile(4, i)

            assertEquals("#" + i, expected, actual)
        }
    }

    companion object {
        private const val SOURCES_COUNT = 4
        private var outputter: XMLOutputter? = null

        @BeforeClass
        @JvmStatic
        fun before() {
            outputter = XMLOutputter()
            val format = Format.getPrettyFormat().setLineSeparator("\n")
            outputter!!.format = format
        }

        @Throws(IOException::class, SAXException::class, ParserConfigurationException::class)
        private fun readSourceXmlFile(version: Int, index: Int): Element {
            return TestUtils.readXmlFile(StudySerializationUtilsTest::class.java, format("version{0}_{1}.xml", version, index))
        }

        @Throws(IOException::class)
        private fun readSourceTextFile(version: Int, index: Int): String {
            return TestUtils.readTextFile(StudySerializationUtilsTest::class.java, format("version{0}_{1}.xml", version, index))
        }
    }
}
